### ISSUE TITLE
`reportUnknownArgumentType` warns even if `object` is accepted

### ISSUE THREAD

**Issue by parched** [microsoft/pyright#3552]
**Describe the bug**
`reportUnknownArgumentType` is produced even if any type is valid (so it shouldn't matter that it's unknown).

**To Reproduce**
```
def use_anything(obj: object):  # if this is `obj: Any` the warning goes away
                                # but I want type checking inside the function
    pass


def check_anything(obj: object):
    match obj:
        case str():
            pass
        case list():
            pass
        case _:
            raise TypeError

    use_anything(obj) # <- the following warning is produced for 'obj'
                      # Argument type is partially unknown
                      #   Argument corresponds to parameter "obj" in function "use_anything"
                      #   Argument type is "str | list[Unknown]"Pyright[reportUnknownArgumentType]
```

**Expected behavior**
I expect no warning

**Comment by erictraut** [comment]
This is expected behavior. The `case list()` statement narrows the type of `obj` to `list[Unknown]`. You'll see the same thing if you use the `isinstance` equivalent:

```python
def check_anything(obj: object):
    if isinstance(obj, str):
        pass
    elif isinstance(obj, list):
        pass
    else:
        raise TypeError

    use_anything(obj)
```

The problem is that using `isinstance` (or `cast list()`) narrows the type to a `list`, but the type argument cannot be determined, so it is `Unknown`.

The `reportUnknownArgumentType` rule is a strict-mode check, and you may need to make accommodations in your code if you want to use it. This is an example. Consider ways to restructure the code or change its behavior to avoid this situation. I'd normally offer a workaround, but I don't know enough about your use case to offer anything useful.

As a last resort, you can use a `# pyright: ignore[reportUnknownArgumentType]` to suppress the error on that line.

**Comment by parched** [comment]
@erictraut but why does the error go away when I change the parameter type to `Any` for `use_anything`? The same narrowing logic applies. If there's a special case here for `Any` shouldn't it apply to `object` too?

My actual code looks like this (here `obj` is a JSON compatible type, I've typed it as `object` but, really, a recursive union would be better if that's possible)
```
def _get_replacements(
    name: str, obj: object, replacer: Callable[[object], _T]
) -> list[tuple[str, _T]]:
    """Return the replacements with as much as can be expanded first"""
    replacements = list[tuple[str, _T]]()
    match obj:
        case str():
            pass
        case float() | int() | bool() | None:
            pass
        case list():
            obj_ = cast(list[object], obj)
            replacements.extend(
                (
                    replacement
                    for i, value in enumerate(obj_)
                    for replacement in _get_replacements(
                        f"{name}[{i}]", value, replacer
                    )
                )
            )
        case dict():
            obj_ = cast(dict[str, object], obj)
            replacements.extend(
                (
                    replacement
                    for key, value in obj_.items()
                    for replacement in _get_replacements(
                        f"{name}.{key}", value, replacer
                    )
                )
            )
        case _:
            raise TypeError("Unsupported JSON type")

    replacement = replacer(obj)  # pyright: ignore [reportUnknownArgumentType]
    # https://github.com/microsoft/pyright/issues/3552
    replacements.append((name, replacement))
    return replacements
```

**Comment by parched** [comment]
I just tried a recursive union for the JSON type and it worked, it also allowed me to get rid of those `cast`s too :).
